### ISSUE TITLE
Review and decide a better way to use CARTO.js functions out of deep-insights-integrations

### ISSUE THREAD

**Issue by xavijam** [CartoDB/cartodb#11846]
Basically, we have open the pandora box with the `export as image` functionality, because we are invoking the "global" variable `vis` provided from CARTO.js in order to use the method `getStaticImageURL `. 

We should find a good way to not start using methods from CARTO.js vis instances as soon as possible. So let's do it before releasing this feature.

cc @javierarce @alonsogarciapablo

**Comment by alonsogarciapablo** [comment]
I can't agree with you more @xavijam. We shouldn't use the global `vis` variable.

I like the way we're doing things now, where we have some sort of façade between BUILDER and deep-insights.js / cartodb.js. Yes, I'm talking about the (in)famous [deep-insights-integrations.js](https://github.com/CartoDB/cartodb/blob/master/lib/assets/core/javascripts/cartodb3/deep-insights-integrations.js). The file is messy and very big but it avoids leaking (almost) all deep-insights.js / cartodb.js objects into BUILDER. I think we can follow a similar approach. It's a bit different case here, because BUILDER usually just invokes some function from deep-insights.js / cartodb.js in response to an event on one of BUILDER models or viceversa. Syntonisation  between both systems is normally one way only. In this case BUILDER needs to ask CartoDB.js for something and work with the response immediately.

Here're some things I can think of:

### 1. CartoDB.js exposes a URL template for the image instead of generating the URL of the image

Instead of generating the URL based on some parameters, `vis.getStaticImageURL` could return something like "http://wadus.com/api/v1/map/static/center/1234567890/{z}/{lat}/{lng}/{width}/{height}.{format}?layerIndex=0,1,2&api_key=0987654321".

```
vis.on('reloaded', function () {
  visDefinitionModel.set('staticImageURLTemplate', vis.getStaticImageURL);
});

// then visDefinitionModel can define `getStaticImageURL`

getStaticImageURL: function (options) {
  options = _.defaults({}, options, {
    zoom: 4,
    lat: 0,
    lng: 0,
    width: 300,
    height: 300,
    format: 'png'
  });
  var url;
  var urlTemplate = this.get('staticImageURLTemplate');
  if (urlTemplate) {
    url = urlTemplate
      .replace('{z}', options.zoom)
      .replace('{lat}', options.lat)
      .replace('{lng}', options.lng)
      .replace('{width}', options.width)
      .replace('{height}', options.height)
      .replace('{format}', options.format);
  }
  return url;
}

// Later, somewhere else...

visDefinitionModel.getStaticImageURL({ zoom: 10 });
```

### 2. We could wrap that function/vis object in a BUILDER object/function


```

var StaticImageGenerator = function (vis) {
  this._vis = vis;
}

StaticImageGenerate.prototype.generateURL = function (options) {
  this._vis.getStaticImageURL(options);
};
var staticImageGenerator = new StaticImageGenerator(vis);

// we can pass that object around and then do something like
staticImageGenerator.generateURL({...});
```

This way, BUILDER doesn't know about `vis` directly, but still `vis` is "leaked" somehow into BUILDER...

Other options? Thoughts?

**Comment by xavijam** [comment]
I really like this option @alonsogarciapablo :+1:. I was thinking of setting zoom and center (also bounds, but they are already set) in the state definition model, with that we will have any data we need from different models without requesting `vis` global model, so 💃 .

**Comment by alonsogarciapablo** [comment]
Which of the two @xavijam?

**Comment by xavijam** [comment]
First, sorry!

**Comment by xavijam** [comment]
Hey, whoever takes this issue, take into account we should PR from this [issue](https://github.com/CartoDB/cartodb/issues/11775), with last export-image changes.

**Comment by nobuti** [comment]
Ready for review. Adding the `don't merge yet` label because it relays on PRs in cartodb.js and deep-insights.js we need to merge before this.

**Comment by nobuti** [comment]
Closed by #11915.
